drivers/hmis/app/graphql/types/forms/form_definition.rb

```diff
@@ -48,6 +49,24 @@ def definition
       )
     end
 
+    def project_matches
+      # Returns one match per project that this form applies to. (See InstanceProjectMatch for match ranking logic.)
+      # This DOES still return a match for a project if the match is overridden by a more specific rule on another form.
+      # For example, if:
+      # - Intake form A has a rule that specifies that it's used for all projects
+      # - Intake form B has a rule that specifies that it's used for all Emergency Shelter projects
+      # - Intake form C has a rule that specifies that it's used for Project X, an ES project
+      # ...then forms A and B would still return a match for Project X.
+      object.instances.
+        active.
+        map(&:project_matches).
+        flatten.
+        sort_by(&:rank).
+        reverse. # lower rank means a better match
+        map { |match| [match.project.id, match] }. # per project ID, the later added (lower ranked) matches overwrite the earlier ones
+        to_h.values
+    end
```

Comment:
Nice!

This looks right. As discussed, the fact that `project_matches` checks against all projects is not performant, so that could be further optimized (ended up with an idea below, which doesn't seem too cumbersome).

We should also attempt to reduce N+1s on project when resolving, which we can do with a preload.

It also seems unnecessary that  `Hmis::Hud::Project.all` is queried multiple times. below is a rough suggestion of passing a `scope` that should reduce that:

```
instance_scope = object.instances

# fetching all projects is slow, but at least it happens only once. preload organization because we are going to resolve it
# `.hmis` scope is necessary to exclude unrelated warehouse projects
project_scope = Hmis::Hud::Project.hmis.preload(:organization)

# preload funders only if any of these rules are funder-based (micro-optimizing)
project_scope = project_scope.preload(:funders) if instance_scope.where.not(funder: nil).exists?

# future optimization: further refine project_scope down to only projects that could be matched by any of the instances

# returns `Hmis::Form::InstanceProjectMatch`
object.instances.active.map { |instance| instance.project_matches(scope: project_scope) }.flatten...etc
```

And here's a rough (untested) pass at how we could reduce the size of the project_scope:
```
    matches_any_project = p_t[:id].in(instances.select { |inst| inst.entity_type == 'Hmis::Hud::Project' }.map(&:entity_id))
    matches_org = o_t[:id].in(instances.select { |inst| inst.entity_type == 'Hmis::Hud::Organization' }.map(&:entity_id))
    matches_funders = f_t[:funder].in(instances.map(&:funder))
    matches_project_types = f_t[:project_type].in(instances.map(&:project_type))

    possible_refined_project_scope = if instances.defaults.exists?
      # there is a default so all projects are fair game
      Hmis::Hud::Project.hmis.all
    else
      # glom together all the projects that might match any of the rules
      # this still would need a second pass with the match class to express the full logic (eg "funder AND project type" match) and to get the rank
      Hmis::Hud::Project.hmis.joins(:organization).left_outer_joins(:funders).where(
        matches_any_project.or(matches_funders).or(matches_project_types).or(matches_org),
      )
    end
```

Reply:
thanks @gigxz this is great, I'd like to add more tests but it seems to be working ok against the existing test!

Reply:
cool! more test cases would be great